### PR TITLE
Translate failure messages of json authentication

### LINKED ISSUE #33168: Should Symfony translate security exception messages?
**Description**  
Symfony includes the translation of Security exception messages ... but they are not used automatically.

For example, if you build a custom Guard-based login mechanism ... `UsernamePasswordJsonAuthenticationListener` can return `return new JsonResponse(['error' => $failed->getMessageKey()], 401);` and you get `Bad credentials.` message in English, regardless of the current application locale.

Is this correct? Look at PHPdoc:

```php
   /**
     * Message key to be used by the translation component.
     *
     * @return string
     */
    public function getMessageKey()
```

So, this is *meant* to be translated. Should we do it?

In the [login form docs](https://symfony.com/doc/current/security/form_login_setup.html) we translate this manually:

```twig
   {% if error %}
        <div class="alert alert-danger">{{ error.messageKey|trans(error.messageData, 'security') }}</div>
    {% endif %}
```

Should this message come translated instead?

### PR DESCRIPTION
**Pull request by malteschlueter**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Resolves #33168
| License       | MIT
| Doc PR        | -

Until now the failure messages of the json authentication were not translated. I'm not sure if it's a bug or a new feature. The changes shouldn't be a BC.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php [hunk 1]
```diff
@@ -143,6 +143,7 @@
                 abstract_arg('options'),
                 service('property_accessor')->nullOnInvalid(),
             ])
+            ->call('setTranslator', [service('translator')->nullOnInvalid()])
```
Comment (wouterj) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php:146]: I think `ignoreOnInvalid()` is better (it results in the setter not being called if the `translator` service is not available)

same below

File: src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php [hunk 2]
```diff
@@ -131,6 +143,11 @@ public function isInteractive(): bool
         return true;
     }
 
+    public function setTranslator(?TranslatorInterface $translator)
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php:146]: if we make the DI config `ignoreOnInvalid`, we can remove nullable here.

**Review by wouterj — COMMENTED** [review]
Thanks for working on this one. It's definitely a feature (so targetting master is a good idea).

I don't think we have to add this to the `UsernamePasswordJsonAuthenticationListener` (those are part of a system that will be removed in Symfony 6). If you can, I think it's also a good idea to add a test for this new feature.

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php (hunk 1 above)
Comment (malteschlueter) [reply src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php]: Good idea :)

File: src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php (hunk 2 above)
Comment (malteschlueter) [reply src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php]: Good idea :)

**Comment by malteschlueter** [comment]
> Thanks for working on this one. It's definitely a feature (so targetting master is a good idea).

My pleasure.

> I don't think we have to add this to the `UsernamePasswordJsonAuthenticationListener` (those are part of a system that will be removed in Symfony 6). If you can, I think it's also a good idea to add a test for this new feature.

I will add a test. Takes some time^^

**Comment by malteschlueter** [comment]
During writing the test i saw some other error messages like "_Invalid username._" [(JsonLoginAuthenticator.php#L167)](https://github.com/symfony/symfony/blob/485e59f43946935c8f07d7aeccbd43e703937b9d/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L167).
Am i blind or is there no translation for it?

What do you think about the other error messages. Should they also translated see "[Invalid JSON.](https://github.com/symfony/symfony/blob/485e59f43946935c8f07d7aeccbd43e703937b9d/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L155)" or "[The key "username" must be provided.](https://github.com/symfony/symfony/blob/485e59f43946935c8f07d7aeccbd43e703937b9d/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L163)"

**Comment by ro0NL** [comment]
i figure we can re-use a bunch of translations from the validator domain.

**Comment by ro0NL** [comment]
no wait, i dont think http exceptions need to be translated at all. Usually they are caught by an ExceptionListener exposing the http status text in prod (while preserving the original message in dev).

In this case im fine with only translating the responses generated here :+1:

**Comment by wouterj** [comment]
Please note that the `$message` constructor is a non-user safe string and the `messageKey` is almost always [hardcoded in the exception class](https://github.com/symfony/symfony/blob/485e59f43946935c8f07d7aeccbd43e703937b9d/src/Symfony/Component/Security/Core/Exception/BadCredentialsException.php#L25-L28). And this message key does exists: https://github.com/symfony/symfony/blob/485e59f43946935c8f07d7aeccbd43e703937b9d/src/Symfony/Component/Security/Core/Resources/translations/security.en.xlf#L17-L20

**Review by ro0NL — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
@malteschlueter looks like the PR already needs a rebase 🙈

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @malteschlueter.

**Comment by malteschlueter** [comment]
Thanks for merging @fabpot. It was a little bit too fast. I missed to add some tests. I've done it here #38044 now. :)

**Pull request merged**